Milk Sum USACO Silver (#5336)
cpinitiative/usaco-guide: solutions/silver/usaco-1326.mdx

---
id: usaco-1326
source: USACO Silver 2023 Open
title: Milk Sum
author: Sachet Abeysinghe
---

[Official Analysis (C++, Java, Python)](https://usaco.org/current/data/sol_prob1_silver_open23.html)

## Explanation

First, observe that the maximum possible amount of milk is generated by assigning the largest multipliers to the cows with the highest milk production values. This is achieved by sorting $a$ in non-decreasing order.

So, for each query, we can try changing the element with value $a_i$ to $v$, and maintaining sorted order by either re-sorting $a$ or utilizing a sorted list. Then, $T$ can be calculated in linear time, leading to $\mathcal{O}(N)$ or $\mathcal{O}(N\log N)$ per query depending on implementation. However, this is too slow with the constraints and only passes the first subtask.

> Let's observe the effect of changing $a_i$ to $v$ when maintaining sorted order and how this affects the value of the original value of $T$. If we consider the sorted version of $a$, then there are two cases: the element with value $a_i$ moves right and elements are shifted left $(v \ge a_i)$, or the element with value $a_i$ moves left and elements are shifted right $(v \lt a_i)$. For the first case, the multipliers of the elements shifted left decrease by 1, so $T$ decreases by the sum of those elements, which can be found using prefix sums. We also have to account for the contributions of the removal of the original element and addition of the new element — we can find their positions using binary search. The latter case can be handled similarly.

Reviewer: can you split this up into multiple paragraphs

## Implementation

**Time Complexity:** $\mathcal{O}((N + Q)\log N)$

<LanguageSection>
<PySection>

```py
import bisect
import itertools

N = int(input())
a = [int(x) for x in input().split()]

# Sorting and prefix sums
sorted_a = sorted(a)
p_sum = [0] + list(itertools.accumulate(sorted_a))

> # Compute original value of T

Reviewer 2: What is T here? Can you be more specific

Author: T is defined in the problem statement

Reviewer: would be better to make it more explicit

---
id: usaco-1326
source: USACO Silver 2023 Open
title: Milk Sum
author: Sachet Abeysinghe
---

[Official Analysis (C++, Java, Python)](https://usaco.org/current/data/sol_prob1_silver_open23.html)

## Explanation

First, observe that the maximum possible amount of milk is generated by assigning the largest multipliers to the cows with the highest milk production values.

> This is achieved by sorting $a$ in nondecreasing order.

Reviewer 3: i think it would be a bit nicer if you defined what these symbols mean or just refer to them by what they are

Reviewer 3: (yes ik they are defined in the problem statement but it just makes for a better reading experience)

Let's observe the effect of changing an element when maintaining sorted order and
how this affects the value of the original value of $T$. Suppose we consider a query that changes $a_i$ to $v$. Let the positions of $a_i$ and $v$ in the sorted version of $a$ be $x$ and $y$, respectively.

If we consider the sorted version of $a$, then there are two cases:
1. If $v \ge a_i$, $a_i$ moves right and the indices $x+1,\ x+2,\ ...,\ y$ are shifted left.
2. If $v < a_i$, $a_i$ moves left and the indices $y,\ y+1,...,x$ are shifted right.

For the first case, the multipliers of the elements shifted left decrease by 1, so $T$ decreases by the sum of those elements, which can be found using prefix sums.
We also have to account for the contributions of the removal of the original element and addition of the new element — we can find their positions using binary search.
The second case can be handled similarly.

## Implementation

**Time Complexity:** $\mathcal{O}((N + Q)\log N)$

<LanguageSection>
<PySection>

```py
import bisect
import itertools

> N = int(input())

Reviewer 3: lowercase n would be better

a = [int(x) for x in input().split()]

# Sorting and prefix sums
sorted_a = sorted(a)

> p_sum = [0] + list(itertools.accumulate(sorted_a))

Reviewer 3: is itertools really necessary here? it saves at most like two lines

Author: idk i just like to be as concise as possible lol ill change it tho

---
id: usaco-1326
source: USACO Silver 2023 Open
title: Milk Sum
author: Sachet Abeysinghe
---

[Official Analysis (C++, Java, Python)](https://usaco.org/current/data/sol_prob1_silver_open23.html)

## Explanation

First, observe that the maximum possible amount of milk is generated by assigning the largest multipliers to the cows with the highest milk production values.

> This is achieved by sorting $a$ from lowest to highest.

Reviewer 2: ```suggestion
This is achieved by sorting $a$ in ascending order.
```

Let's observe the effect of changing $a_i$ to $v$ when maintaining sorted order and
how this affects the value of the original value of $T$.

Let the position of $v$ in $a$ be $j$.
If we have $a$ be sorted already, then there are two cases:
1. If $v \ge a_i$, $a_i$ moves right and the indices $i+1,\ i+2,\ ...,\ j$ are shifted left.

> 2. If $v < a_i$, $a_i$ moves left and $j,\ j+1,...,i$ are shifted right.

Author: The original position of a_i before it's changed to v

> 2. If $v < a_i$, $a_i$ moves left and $j,\ j+1,...,i$ are shifted right.

Author: 😭did u read the problem, i is the position of a_i in the sorted version (not the original version as described in the problem) and j is the position of v in the sorted version

> 2. If $v < a_i$, $a_i$ moves left and $j,\ j+1,...,i$ are shifted right.

Reviewer: > i is the position of a_i in the sorted version

well yeah, but what's special about this index?

> 2. If $v < a_i$, $a_i$ moves left and $j,\ j+1,...,i$ are shifted right.

Reviewer: ok apparently they're specified by the query
you should probably make that clear
"say we're considering a query that changes [x] to [y]"
